### PR TITLE
Better handle large numbers (u128/i128) in deserialization

### PR DESCRIPTION
**Pull request by YDX-2147483647**
Data sources may contain large integers that cannot be represented as `int` in typst.
When decoding large integers in json/yaml/cbor, typst will/should approximate them to `float`.
This behaviour has been explicitly stated in the document, for example, [YAML Function – Typst Documentation](https://typst.app/docs/reference/data-loading/yaml/):

> Be aware that integers larger than $2^{63}-1$ will be converted to floating point numbers, which may give an approximative value.

## Before

However in typst 0.13.1, it only applies to numbers less than `u64::MAX`. For even larger integers in yaml and cbor, an error will be thrown.

| Format | $2^{63} - 1$ | $2^{63}$                | $-2^{63}-1$                      | $2^{63} \times 10$               |
| ------ | ------------ | ----------------------- | -------------------------------- | -------------------------------- |
| JSON   | int          | float                   | float                            | float                            |
| YAML   | int          | float                   | **Error: i128 is not a typst value** | **Error: u128 is not a typst value** |
| CBOR   | int          | float                   | **Error: i128 is not a typst value** | **Error: u128 is not a typst value** |
| TOML | int | Error: number too large | Error: number too small | Error: number too large |

<details><summary>Test code</summary>

```typst
#assert.eq(type(json(bytes("9223372036854775807"))), int)
#assert.eq(type(json(bytes("9223372036854775808"))), float)
#assert.eq(type(json(bytes("-9223372036854775809"))), float)
#assert.eq(type(json(bytes("92233720368547758080"))), float)


#assert.eq(type(yaml(bytes("9223372036854775807"))), int)
#assert.eq(type(yaml(bytes("9223372036854775808"))), float)
// [Error] Failed to parse YAML (invalid type: integer -9223372036854775809 as i128, expected a typst value)
// #yaml(bytes("-9223372036854775809"))
// [Error] Failed to parse YAML (invalid type: integer 92233720368547758080 as u128, expected a typst value)
// #yaml(bytes("92233720368547758080"))

#assert.eq(type(toml(bytes("a = 9223372036854775807")).a), int)
// [Error] Failed to parse TOML (number too small to fit in target type at line 1 column 5)
// #toml(bytes("a = -9223372036854775809"))
// [Error] Failed to parse TOML (number too large to fit in target type at line 1 column 5)
// #toml(bytes("a = 9223372036854775808"))
// [Error] Failed to parse TOML (number too large to fit in target type at line 1 column 5)
// #toml(bytes("a = 92233720368547758080"))

#assert.eq(type(cbor(bytes((27, 127, 255, 255, 255, 255, 255, 255, 255)))), int)
#assert.eq(type(cbor(bytes((27, 128, 0, 0, 0, 0, 0, 0, 0)))), float)
// [Error] Failed to parse CBOR (Semantic(None, "invalid type: integer -9223372036854775809 as i128, expected a typst value")) 
// #cbor(bytes((59, 128, 0, 0, 0, 0, 0, 0, 0)))
// [Error] Failed to parse CBOR (Semantic(None, "invalid type: integer 92233720368547758080 as u128, expected a typst value")) 
// #cbor(bytes((194, 73, 5, 0, 0, 0, 0, 0, 0, 0, 0)))
```

</details> 

The cause is that `Deserialize` implementations may output `i128`/`u128`, but typst refuses to convert them to `Value`.
(Decoding JSON works as expected because `serde_json` has a special treatment in [`Number::as_f64`](https://docs.rs/serde_json/1.0.138/src/serde_json/number.rs.html#162-171).)

## After
This PR will add `u128`/`i128` support.

| Format | $2^{63} - 1$ | $2^{63}$                | $-2^{63}-1$                      | $2^{63} \times 10$               |
| ------ | ------------ | ----------------------- | -------------------------------- | -------------------------------- |
| JSON   | int          | float                   | float                            | float                            |
| YAML   | int          | float                   | **float** | **float** |
| CBOR   | int          | float                   | **float** | **float** |
| TOML | int | Error: number too large | Error: number too small | Error: number too large |

Continues #3791
Relates to https://github.com/typst/typst/pull/6743#discussion_r2306827806

### TOML

The [toml spec](https://toml.io/en/v1.0.0#integer) says:
> Arbitrary 64-bit signed integers (from −2^63 to 2^63−1) should be accepted and handled losslessly. If an integer cannot be represented losslessly, **an error must be thrown.**

### Test cases

I added four test cases: `{json,yaml}-decode-number` and `{cbor,toml}-decode-integer`.
Their names are not consistent, because integer and floats are distinct in toml and cbor, but unified as _numbers_ in json and yaml.

In addition, the test case for cbor is generated from the following rust script. I can move it to typst-dev-assets if appropriate.

<details><summary>The script for the CBOR test case</summary>

`src/main.rs`:

```rust
use serde::Serialize;

#[derive(Serialize, Default)]
struct Data {
    representable: RepresentableInteger,
    large: LargeInteger,
}

#[derive(Serialize)]
struct RepresentableInteger {
    i64_max: i64,
    i64_min: i64,
}
impl Default for RepresentableInteger {
    fn default() -> Self {
        Self {
            i64_max: i64::MAX,
            i64_min: i64::MIN,
        }
    }
}

#[derive(Serialize)]
struct LargeInteger {
    i64_max_plus_one: u64,
    i64_min_minus_one: i128,
    u64_max: u64,
    u64_max_plus_one: i128,
    i128_max: i128,
    i128_min: i128,
    u128_max: u128,
}

impl Default for LargeInteger {
    fn default() -> Self {
        Self {
            i64_max_plus_one: (i64::MAX as u64) + 1,
            i64_min_minus_one: (i64::MIN as i128) - 1,
            u64_max: u64::MAX,
            u64_max_plus_one: (u64::MAX as i128) + 1,
            i128_max: i128::MAX,
            i128_min: i128::MIN,
            u128_max: u128::MAX,
        }
    }
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let data = Data::default();

    let mut writer = Vec::new();
    ciborium::into_writer(&data, &mut writer)?;

    println!("{writer:?}");
    Ok(())
}
```

`Cargo.toml`:

```toml
[package]
name = "test-cbor"
version = "0.1.0"
edition = "2024"

[dependencies]
ciborium = "0.2.2"
serde = { version = "1.0.219", features = ["derive"] }
```

</details> 

Finally, edge cases in serialization like `strong[Lorem]` may also be added to `tests/suite/loading/edge-case.typ` in the future. (#6743)

### CODE REVIEW

File: tests/suite/loading/edge-case.typ [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  i64-min-minus-one: "-9223372036854775809",
+  u64-max: "18446744073709551615",
+  u64-max-plus-one: "18446744073709551616",
+  i128-max: "170141183460469231731687303715884105727",
+  i128-min: "-170141183460469231731687303715884105728",
+  u128-max: "340282366920938463463374607431768211455",
+)
+
+// TODO: The following is generated from the rust script in PR description.
+// I can move the script to typst-dev-assets if appropriate.
```
Comment (laurmaedje) [inline tests/suite/loading/edge-case.typ]: No need. This is fine.

**Comment by YDX-2147483647** [comment]
> remove the TODO

OK! I left a link to this PR. Hope it will help future historians.
